src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp

```diff
@@ -121,19 +121,19 @@ ShenandoahGenerationalHeap::TransferResult ShenandoahGenerationalHeap::balance_g
   shenandoah_assert_heaplocked_or_safepoint();
 
   ShenandoahOldGeneration* old_gen = old_generation();
-  const size_t old_region_surplus = old_gen->get_region_surplus();
-  const size_t old_region_deficit = old_gen->get_region_deficit();
-  old_gen->set_region_surplus(0);
-  old_gen->set_region_deficit(0);
+  const ssize_t old_region_balance = old_gen->get_region_balance();
+  old_gen->set_region_balance(0);
 
-  if (old_region_surplus) {
-    bool success = generation_sizer()->transfer_to_young(old_region_surplus);
+  if (old_region_balance > 0) {
+    const auto old_region_surplus = checked_cast<size_t>(old_region_balance);
+    const bool success = generation_sizer()->transfer_to_young(old_region_surplus);
     return TransferResult {
       success, old_region_surplus, "young"
     };
   }
 
-  if (old_region_deficit) {
+  if (old_region_balance < 0) {
+    const auto old_region_deficit = checked_cast<size_t>(-old_region_balance);
```

Comment:
Like I remarked elsewhere, the code in the if > 0 and if < 0 branches (or if you will, the if and else branches) are largely isomorphic, and can probably be folded together by calling a generic `transfer_from_old(ssize_t num_regions)` on ShenandoahGenerationSizer. The latter new method would be a folding together of the old methods `transfer_to_young()` and `transfer_to_old()`, and would return the `TransferResult` that this method `balance_generations` returns to its caller.